### PR TITLE
Use actions/setup-go@v5 in github actions

### PR DESCRIPTION
**Pull request by kke**
## Description

Use actions/setup-go@v5 instead of v3.

Closes #3795

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## How Has This Been Tested?

- [ ] Manual test
- [ ] Auto test added

## Checklist:

- [ ] My code follows the style [guidelines](https://github.com/k0sproject/k0s/blob/main/docs/contributors/overview.md) of this project 
- [ ] My commit messages are [signed-off](https://github.com/k0sproject/k0s/blob/main/docs/contributors/github_workflow.md)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by twz123** [comment]
Maybe better use `with: {cache: false}` for now? We'd need to inspect our own caching and replace that. Since there's actions/cache/save and actions/cache/restore available nowadays, there might be some potential to completely rework this.

**Comment by kke** [comment]
I think we're manually caching just the docker-builder deps like:

```yaml
      - name: Cache GOMODCACHE
        uses: actions/cache@v3
        with:
          key: unittests-k0s-linux-amd64-gomodcache-${{ hashFiles('go.sum') }}
          restore-keys: |
            build-k0s-linux-amd64-gomodcache-${{ hashFiles('go.sum') }}
          path: |
            build/cache/go/mod
```

There was one which cached `~/go/pkg/mod` which I now removed.

File: .github/workflows/autopilot-branch-to-latest.yaml [hunk 1]
```diff
@@ -191,9 +191,12 @@ jobs:
       run: .github/workflows/prepare-build-env.sh
 
     - name: Set up Go
-      uses: actions/setup-go@v3
+      uses: actions/setup-go@v5
       with:
-        go-version: ${{ env.GO_VERSION }}
+        go-version-file: go.mod
```
[the remark this answers is not in the document]
Comment (kke) [reply .github/workflows/autopilot-branch-to-latest.yaml:196]: Maybe the makefiles need to get it from there too.

I have some doubts about the version selection in setup-go and the `check-latest: true` setting.

File: .github/workflows/go.yml [hunk 2]
```diff
@@ -224,9 +218,12 @@ jobs:
         run: .github/workflows/prepare-docker-ipv6.sh
 
       - name: Set up Go
-        uses: actions/setup-go@v3
+        uses: actions/setup-go@v5
```
[the remark this answers is not in the document]
Comment (kke) [reply .github/workflows/go.yml:221]: There's a `cache: false` setting that could be used

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -68,9 +68,12 @@ jobs:
           VERSION: ${{ needs.release.outputs.tag_name }}
 
       - name: Set up Go for smoke tests
-        uses: actions/setup-go@v3
+        uses: actions/setup-go@v5
```
[the remark this answers is not in the document]
Comment (kke) [reply .github/workflows/release.yml:71]: Would that potentially confuse SBOM?

File: .github/workflows/publish-docs.yml [hunk 4]
```diff
@@ -27,10 +27,13 @@ jobs:
...
-      - name: Set up Go ${{ env.GO_VERSION }}
-        uses: actions/setup-go@v3
+      - name: Set up Go
+        uses: actions/setup-go@v5
         with:
-          go-version: ${{ env.GO_VERSION }}
+          go-version-file: go.mod
+          cache-dependency-path: |
+            go.sum
+            hack/tool/go.sum
```
Comment (twz123) [inline .github/workflows/publish-docs.yml:36]: The go toolchain here is solely required to compile k0s_sort, so there's no need to cache this.

This is also the biggest concern that I have about using setup-go's builtin caching. It's made for the simple case that there's a single workflow that installs go once and always compiles the same stuff. Perfectly reasonable for most of the go projects out there. The k0s workflows are different, and compile different parts of the project in different workflows. Yet, setup-go will always use the same cache for all of those. Here, we're not even compiling anything from k0s's source tree.

File: .github/workflows/publish-docs.yml [hunk 5]
```diff
@@ -27,10 +27,13 @@ jobs:
           cache: pip
           cache-dependency-path: docs/requirements.txt
 
-      - name: Set up Go ${{ env.GO_VERSION }}
-        uses: actions/setup-go@v3
+      - name: Set up Go
+        uses: actions/setup-go@v5
```
Comment (twz123) [inline .github/workflows/publish-docs.yml:31]: We could replace this step with a curl/wget if k0s_sort would be available as precompiled executable.

File: .github/workflows/go.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (twz123) [reply .github/workflows/go.yml:221]: That's what I've asked for in https://github.com/k0sproject/k0s/pull/3914#issuecomment-1884798118 :sweat_smile:

File: .github/workflows/autopilot-branch-to-latest.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (twz123) [reply .github/workflows/autopilot-branch-to-latest.yaml:196]: Are we using check-latest anywhere? I don't find it, and the default seems to be false.

File: .github/workflows/release.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (twz123) [reply .github/workflows/release.yml:71]: Not sure I follow. What do you have in mind?

File: .github/workflows/autopilot-branch-to-latest.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (kke) [reply .github/workflows/autopilot-branch-to-latest.yaml:196]: No. It's used in k0sctl and for some reason I've seen it use 1.20 even though go.mod says 1.21.

File: .github/workflows/release.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (twz123) [reply .github/workflows/release.yml:71]: Oh, I see.

File: hack/tools/gen-matrix.sh [hunk 6]
```diff
@@ -14,11 +14,11 @@ list_k0s_releases() {
 }
 
 k0s_sort() {
-  go run github.com/k0sproject/version/cmd/k0s_sort@v0.4.2
+  go run github.com/k0sproject/version/cmd/k0s_sort@v0.5.0 -l
```
Comment (twz123) [inline hack/tools/gen-matrix.sh:17]: v0.5.0 won't work as-is for this script: k0sproject/version#10

**Pull request closed without merging**